programs/vote/src/vote_state.rs

```diff
@@ -489,6 +554,43 @@ pub fn process_vote(
     vote_account.set_state(&vote_state)
 }
 
+pub fn slash_state(
+    vote_account: &mut KeyedAccount,
+    slot_hashes: &[SlotHash],
+    tx: &Transaction,
+) -> Result<(), InstructionError> {
+    let mut vote_state: VoteState = vote_account.state()?;
+    //verify the tx is valid
+    if !tx.verify_refs() {
+        return Err(InvalidSlashTransaction);
+    }
+    //verify transaction signature
+    tx.verify()?;
+    let signers: HashSet<_> = tx
+        .message
+        .account_keys
+        .iter()
+        .take(tx.signatures.len())
+        .collect();
+    //find vote instruction
+    for (i, ix) in tx.message.instructions.iter().enumerate() {
+        if tx.instruction_program(i) != program_id {
+            continue;
+        }
+        if let VoteInstruction::Vote(slot) = limited_deserialize(ix.data) {
+            //verify that account matches mine
+            if tx.key(i, 0) != vote_account.key {
+                continue;
+            }
+            //verify that the transaction was signed by the expected signers
+            vote_state::verify_authorized_signer(&vote_state.authorized_voter, &signers)?;
```

Comment:
I think this means you can escape a slashing if you update your authorized voter key right after trying to cheat?  we might need to limit how often authorized_voter can change so we can keep a list of them

Reply:
yea, key rotation and slashing is a huge Pia.

Reply:
#7555